Memory usage in docker container creation - Docs

Building MozDef on my 2GB/no swap vm fails. After looking into this I think the build process take just under 4GB of memory. This peaked during the meteor container.
a) Can someone confirm my observation.
b) This should be noted somewhere.

```
Step 22/24 : RUN if [ "${METEOR_BUILD}" = "YES" ]; then         cd /opt/mozdef/envs/mozdef/meteor &&         meteor npm install &&         meteor build --server localhost:3002 --directory /opt/mozdef/envs/meteor/mozdef &&         cp -r /opt/mozdef/envs/mozdef/meteor/node_modules /opt/mozdef/envs/meteor/mozdef/node_modules &&        cd /opt/mozdef/envs/meteor/mozdef/bundle/programs/server &&         npm install ;  fi
 ---> Running in ce04a61ebf8b

> bcrypt@3.0.1 install /opt/mozdef/envs/mozdef/meteor/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Using needle for node-pre-gyp https download
node-pre-gyp WARN Tried to download(404): https://github.com/kelektiv/node.bcrypt.js/releases/download/v3.0.1/bcrypt_lib-v3.0.1-node-v57-linux-x64-glibc.tar.gz
node-pre-gyp WARN Pre-built binaries not found for bcrypt@3.0.1 and node@8.11.4 (node-v57 ABI, glibc) (falling back to source compile with node-gyp)
make: Entering directory `/opt/mozdef/envs/mozdef/meteor/node_modules/bcrypt/build'
  CXX(target) Release/obj.target/bcrypt_lib/src/blowfish.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node
  COPY Release/bcrypt_lib.node
  COPY /opt/mozdef/envs/mozdef/meteor/node_modules/bcrypt/lib/binding/bcrypt_lib.node
  TOUCH Release/obj.target/action_after_build.stamp
make: Leaving directory `/opt/mozdef/envs/mozdef/meteor/node_modules/bcrypt/build'
added 157 packages from 208 contributors and audited 711 packages in 8.299s
found 0 vulnerabilities


<--- Last few GCs --->

[174:0x228bd00]   518731 ms: Scavenge 1485.7 (1591.1) -> 1479.6 (1594.6) MB, 20.7 / 0.0 ms  allocation failure
[174:0x228bd00]   518772 ms: Scavenge 1489.3 (1594.6) -> 1480.0 (1599.6) MB, 19.1 / 0.0 ms  allocation failure
[174:0x228bd00]   518852 ms: Scavenge 1494.3 (1599.6) -> 1484.2 (1600.6) MB, 16.9 / 0.0 ms  allocation failure
[174:0x228bd00]   518937 ms: Scavenge 1495.7 (1600.6) -> 1487.9 (1605.6) MB, 38.2 / 2.5 ms  allocation failure


<--- JS stacktrace --->
Cannot get stack trace in GC.
FATAL ERROR: NewSpace::Rebalance Allocation failed - process out of memory
 1: node::Abort() [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 2: 0x8c21ec [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 5: 0xa2f63b [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 6: v8::internal::MarkCompactCollector::Evacuate() [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 7: v8::internal::MarkCompactCollector::CollectGarbage() [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 8: v8::internal::Heap::MarkCompact() [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
 9: v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
10: v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
11: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
12: v8::internal::Factory::NewStringFromUtf8(v8::internal::Vector<char const>, v8::internal::PretenureFlag) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
13: v8::String::NewFromUtf8(v8::Isolate*, char const*, v8::NewStringType, int) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
14: node::StringBytes::Encode(v8::Isolate*, char const*, unsigned long, node::encoding, v8::Local<v8::Value>*) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
15: 0x8e2526 [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
16: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
17: 0xad712c [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
18: v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) [/opt/mozdef/.meteor/packages/meteor-tool/.1.8.0.1gdnmok.4nus++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node]
19: 0x343efdf842fd
/bin/sh: line 1:   174 Aborted                 meteor build --server localhost:3002 --directory /opt/mozdef/envs/meteor/mozdef
ERROR: Service 'meteor' failed to build: The command '/bin/sh -c if [ "${METEOR_BUILD}" = "YES" ]; then         cd /opt/mozdef/envs/mozdef/meteor &&         meteor npm install &&         meteor build --server localhost:3002 --directory /opt/mozdef/envs/meteor/mozdef &&         cp -r /opt/mozdef/envs/mozdef/meteor/node_modules /opt/mozdef/envs/meteor/mozdef/node_modules &&        cd /opt/mozdef/envs/meteor/mozdef/bundle/programs/server &&         npm install ;  fi' returned a non-zero code: 134
make: *** [build] Error 1
```

--- a/docs/source/demo.rst
+++ b/docs/source/demo.rst
@@ -3,16 +3,23 @@ Demo Instance
 
 Mozilla used to maintain a demo instance of MozDef, however it's currently offline. Best bet for demo is to clone the repo and use the docker containers to stand up a local instance for yourself. There are some scripts in the https://github.com/mozilla/MozDef/tree/master/examples/demo folder to send sample events into a local docker instance that will generate alerts, attackers, etc and give you a feel for the flow.
 
-Here's how to make MozDef go using the provided docker compose files:
+First; set up your docker environment with some tweaks to avoid some common pitfalls: 
+
+1) Allocate it at least 4GB of memory
+2) Use the aufs filesystem driver ( to avoid issues unpacking tar files on overlayfs)
+
+.. image:: https://user-images.githubusercontent.com/566889/47741098-ac306e80-dc36-11e8-88cb-4ba3f1458028.png
+  :width: 40px
+  :align: center
+  :height: 100px
+
+
+Once you've done that, here's how to make MozDef go using the provided docker compose files:
 
 1) Pull the repo: git clone https://github.com/mozilla/MozDef.git
-2) Build the docker containers:
+2) Run the containers:
 
-  docker-compose -f docker/compose/docker-compose.yml -f docker/compose/docker-compose-rebuild.yml -p mozdef build
-
-3) Run the containers:
-
-  docker-compose -f docker/compose/docker-compose.yml -f docker/compose/docker-compose-rebuild.yml -p mozdef up
+  docker-compose -f docker/compose/docker-compose.yml  -p mozdef up
 
 4) Firefox yourself to http://localhost to see the main UI (when the container starts)
 5) Login using a locally created account (click login, create and choose a username/password)

